virtcontainers/factory/template/template.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2018 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// template implements base vm factory with vm templating.
+
+package template
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+	"time"
+
+	vc "github.com/kata-containers/runtime/virtcontainers"
+	"github.com/kata-containers/runtime/virtcontainers/factory/base"
+	"github.com/kata-containers/runtime/virtcontainers/factory/direct"
+)
+
+type template struct {
+	statePath string
+	config    vc.VMConfig
+}
+
+// Fetch finds and returns a pre-built template factory.
+// TODO: save template metadata and fetch from storage.
+func Fetch(config vc.VMConfig) (base.FactoryBase, error) {
+	statePath := vc.RunVMStoragePath + "/template"
+	t := &template{statePath, config}
+
+	err := t.checkTemplateVM()
+	if err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// New creates a new VM template factory.
+func New(config vc.VMConfig) base.FactoryBase {
+	statePath := vc.RunVMStoragePath + "/template"
+	t := &template{statePath, config}
+
+	err := t.createTemplateVM()
+	if err != nil {
+		// fallback to direct factory if template is not supported.
+		return direct.New(config)
+	}
+
+	return t
+}
+
+// Config returns template factory's configuration.
+func (t *template) Config() vc.VMConfig {
+	return t.config
+}
+
+// GetBaseVM creates a new paused VM from the template VM.
+func (t *template) GetBaseVM() (*vc.VM, error) {
+	return t.createFromTemplateVM()
+}
+
+// CloseFactory cleans up the template VM.
+func (t *template) CloseFactory() {
+	syscall.Unmount(t.statePath, 0)
+	os.RemoveAll(t.statePath)
+}
+
+func (t *template) createTemplateVM() error {
+	// create and mount tmpfs for the shared memory file
+	err := os.MkdirAll(t.statePath, 0700)
+	if err != nil {
+		return err
+	}
+	flags := uintptr(syscall.MS_NOSUID | syscall.MS_NODEV)
+	opts := fmt.Sprintf("size=%dM", t.config.HypervisorConfig.DefaultMemSz+8)
```

Comment:
Why `+8`?

Reply:
An experience value that a vm device state file is smaller than 8MB, usually it is about 3MB~ and we need to align it by 8.